### PR TITLE
feat(examples): add parimutuel payout math library

### PR DESCRIPTION
**Pull request by zardozmonopoly**
Small p/ package for pool-based payout math, meant to be reusable by any realm that needs it (prediction markets, sports betting, anything pool-based) instead of everyone rewriting the same formula inline.

Three functions:
- CalculatePayout -- proportional payout from a winning pool, drawn from the full pool
- CalculatePayoutWithRake -- same thing but with an optional commission (rake, in basis points) taken off the top first
- ImpliedProbabilityBps -- market-implied probability of an outcome based on its share of the pool

Probability is returned as integer basis points rather than float64 on purpose -- floating point math is not guaranteed to be identical across platforms, and this needs to produce the same result on every validator.

Tests cover the normal paths plus the edge cases (zero pool, bet bigger than the winning pool, rake that eats the whole pool, etc). Worth mentioning: the full-rake test actually caught a real bug while writing this -- 100% rake left netPool at 0, which tripped CalculatePayouts own validation and panicked instead of just returning 0. Fixed by special-casing an exhausted pool in CalculatePayoutWithRake.

Happy to adjust naming or add more cases if anything looks off.

### CODE REVIEW

**Review by davd-gzl — COMMENTED** [review]
[AI bot - Automatic review]

Automated technical pass: does the code build, run, and behave as described. No design or scope judgement, and no merge verdict. Posted to give a human reviewer a head start.

Verified on e3b7a7934: a Go port of the three functions returns the same wrapped values the gno run produces, so the overflow is deterministic across validators rather than a consensus split. An exhaustive sweep of three-way splits for every total up to 120 found no case where the payouts sum above the pool, and dust never exceeded 2 units.

Full review: https://github.com/samouraiworld/gno-agent-workspace/blob/main/reviews/pr/5xxx/5975-parimutuel-payout-math/1-e3b7a7934/review_claude-opus-4-8_davd-gzl.md

File: examples/gno.land/p/demo/parimutuel/parimutuel.gno [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+// mulDiv computes (a * b) / c. Amounts are assumed to fit comfortably
+// within int64 for realistic ugnot-denominated pools; for pools that
+// could approach 2^63, callers should pre-scale inputs or use a
+// bignum-backed alternative, which this package does not provide.
+func mulDiv(a, b, c int64) int64 {
+	if c == 0 {
+		panic("parimutuel: division by zero")
+	}
+	return (a * b) / c
+}
```
Comment (davd-gzl) [inline examples/gno.land/p/demo/parimutuel/parimutuel.gno:104]: Critical: a winner-takes-all payout wraps to negative at a pool of 3037000500 ugnot, about 3037 GNOT, and returns with no panic; `ImpliedProbabilityBps` wraps the same way at an outcome pool of 922337203685478. The [doc comment](https://github.com/gnolang/gno/blob/e3b7a7934/examples/gno.land/p/demo/parimutuel/parimutuel.gno#L68-L71) bounds the individual amounts, but the `a*b` product is the real limit. [`overflow.Mul64p`](https://github.com/gnolang/gno/blob/e3b7a7934/gnovm/stdlibs/math/overflow/overflow_generated.gno#L582-L589) is already used for exactly this in [`p/demo/tokens/grc20`](https://github.com/gnolang/gno/blob/e3b7a7934/examples/gno.land/p/demo/tokens/grc20/token.gno#L193-L194).

<details><summary>repro</summary>

```bash
# from a local clone of gnolang/gno:
gh pr checkout 5975 -R gnolang/gno
cat > examples/gno.land/p/demo/parimutuel/overflow_test.gno <<'EOF'
package parimutuel

import "testing"

func TestPayoutDoesNotWrapAtLargePools(t *testing.T) {
	const pool = int64(4_000_000_000) // 4000 GNOT, in ugnot
	if got := CalculatePayout(pool, pool, pool); got != pool {
		t.Fatalf("winner takes all: got %d, want %d", got, pool)
	}
}

func TestImpliedProbabilityDoesNotWrapAtLargePools(t *testing.T) {
	const pool = int64(1_000_000_000_000_000)
	if got := ImpliedProbabilityBps(pool, pool); got != 10000 {
		t.Fatalf("whole pool on one outcome: got %d bps, want 10000", got)
	}
}
EOF
cd examples && go run ../gnovm/cmd/gno test ./gno.land/p/demo/parimutuel; cd ..
rm examples/gno.land/p/demo/parimutuel/overflow_test.gno
```

```
--- FAIL: TestPayoutDoesNotWrapAtLargePools (0.00s)
winner takes all: got -611686018, want 4000000000
--- FAIL: TestImpliedProbabilityDoesNotWrapAtLargePools (0.00s)
whole pool on one outcome: got -8446 bps, want 10000
FAIL    ./gno.land/p/demo/parimutuel
```
</details>

File: examples/gno.land/p/demo/parimutuel/parimutuel.gno [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+// assume a non-degenerate pool.
+func CalculatePayoutWithRake(userBet, totalPool, winningPool, rakeBps int64) int64 {
+	if rakeBps < 0 || rakeBps > 10000 {
+		panic("parimutuel: rakeBps must be between 0 and 10000")
+	}
+	netPool := totalPool - mulDiv(totalPool, rakeBps, 10000)
+	if netPool <= 0 {
+		return 0
+	}
+	return CalculatePayout(userBet, netPool, winningPool)
```
Comment (davd-gzl) [inline examples/gno.land/p/demo/parimutuel/parimutuel.gno]: Critical: any outcome holding more than `10000-rakeBps` of the pool panics instead of paying out, so the market cannot be settled. `netPool` goes in as `totalPool`, so [the `winningPool cannot exceed totalPool` guard](https://github.com/gnolang/gno/blob/e3b7a7934/examples/gno.land/p/demo/parimutuel/parimutuel.gno#L15-L17) compares a pre-rake quantity against a post-rake one. With a 5% rake on 1000, a winning pool of 951 panics, and [the winner-takes-all case](https://github.com/gnolang/gno/blob/e3b7a7934/examples/gno.land/p/demo/parimutuel/parimutuel_test.gno#L5-L10) panics under any nonzero rake.

<details><summary>repro</summary>

```bash
# from a local clone of gnolang/gno:
gh pr checkout 5975 -R gnolang/gno
cat > examples/gno.land/p/demo/parimutuel/rake_test.gno <<'EOF'
package parimutuel

import "testing"

func TestRakeWithDominantWinningPool(t *testing.T) {
	// 5% rake on a pool of 1000 leaves 950; 960 of the 1000 sits on the winner.
	if got := CalculatePayoutWithRake(960, 1000, 960, 500); got != 950 {
		t.Fatalf("got %d, want 950", got)
	}
}
EOF
cd examples && go run ../gnovm/cmd/gno test ./gno.land/p/demo/parimutuel; cd ..
rm examples/gno.land/p/demo/parimutuel/rake_test.gno
```

```
panic: parimutuel: winningPool cannot exceed totalPool
Stacktrace:
CalculatePayout<VPBlock(3,0)>(userBet,netPool,winningPool)
    gno.land/p/demo/parimutuel/parimutuel.gno:16
CalculatePayoutWithRake<VPBlock(4,1)>(960,1000,960,500)
    gno.land/p/demo/parimutuel/parimutuel.gno:40
```
</details>

File: examples/gno.land/p/demo/parimutuel/parimutuel.gno [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+// returns 0 rather than delegating to CalculatePayout, whose invariants
+// assume a non-degenerate pool.
+func CalculatePayoutWithRake(userBet, totalPool, winningPool, rakeBps int64) int64 {
+	if rakeBps < 0 || rakeBps > 10000 {
+		panic("parimutuel: rakeBps must be between 0 and 10000")
+	}
+	netPool := totalPool - mulDiv(totalPool, rakeBps, 10000)
+	if netPool <= 0 {
+		return 0
+	}
```
Comment (davd-gzl) [inline examples/gno.land/p/demo/parimutuel/parimutuel.gno]: A negative `totalPool` returns 0 instead of panicking: `netPool` is computed before any validation, so `CalculatePayoutWithRake(100, -1000, 400, 0)` short-circuits at the `netPool <= 0` check while [`CalculatePayout` rejects the same input](https://github.com/gnolang/gno/blob/e3b7a7934/examples/gno.land/p/demo/parimutuel/parimutuel.gno#L11). A caller with the sign wrong gets silent zeros everywhere.

<details><summary>repro</summary>

```bash
# from a local clone of gnolang/gno:
gh pr checkout 5975 -R gnolang/gno
cat > examples/gno.land/p/demo/parimutuel/neg_test.gno <<'EOF'
package parimutuel

import "testing"

func TestNegativeTotalPoolRejected(t *testing.T) {
	defer func() {
		if recover() == nil {
			t.Fatal("negative totalPool accepted")
		}
	}()
	t.Logf("CalculatePayoutWithRake(100, -1000, 400, 0) = %d", CalculatePayoutWithRake(100, -1000, 400, 0))
}
EOF
cd examples && go run ../gnovm/cmd/gno test ./gno.land/p/demo/parimutuel; cd ..
rm examples/gno.land/p/demo/parimutuel/neg_test.gno
```

```
--- FAIL: TestNegativeTotalPoolRejected (0.00s)
CalculatePayoutWithRake(100, -1000, 400, 0) = 0
negative totalPool accepted
FAIL    ./gno.land/p/demo/parimutuel
```
</details>

File: examples/gno.land/p/demo/parimutuel/parimutuel_test.gno [hunk 4]
```diff
@@ -0,0 +1,109 @@
+package parimutuel
+
+import "testing"
+
+func TestCalculatePayoutSingleWinnerTakesAll(t *testing.T) {
+	payout := CalculatePayout(1000, 1000, 1000)
+	if payout != 1000 {
+		t.Fatalf("expected 1000, got %d", payout)
+	}
+}
```
Comment (davd-gzl) [inline examples/gno.land/p/demo/parimutuel/parimutuel_test.gno:10]: Missing test: no case uses a pool large enough for the product in `mulDiv` to leave int64, or a winning pool above `10000-rakeBps` of the total. The largest pool in the suite is 1000, six orders of magnitude below the wrap threshold, and the largest raked `winningPool` is 40% of the total.

<details><summary>test cases</summary>

```go
func TestPayoutDoesNotWrapAtLargePools(t *testing.T) {
	const pool = int64(4_000_000_000) // 4000 GNOT, in ugnot

	if got := CalculatePayout(pool, pool, pool); got != pool {
		t.Fatalf("winner takes all: got %d, want %d", got, pool)
	}
}

func TestImpliedProbabilityDoesNotWrapAtLargePools(t *testing.T) {
	const pool = int64(1_000_000_000_000_000) // 1e9 GNOT, in ugnot

	if got := ImpliedProbabilityBps(pool, pool); got != 10000 {
		t.Fatalf("whole pool on one outcome: got %d bps, want 10000", got)
	}
	if got := ImpliedProbabilityBps(pool/4, pool); got != 2500 {
		t.Fatalf("quarter pool on one outcome: got %d bps, want 2500", got)
	}
}

func TestRakeWithEveryoneOnTheWinner(t *testing.T) {
	// 1% rake, all 1000 staked on the outcome that wins.
	if got := CalculatePayoutWithRake(1000, 1000, 1000, 100); got != 990 {
		t.Fatalf("got %d, want 990", got)
	}
}

func TestRakeWithDominantWinningPool(t *testing.T) {
	// 5% rake leaves 950; a winning pool of 960 is still a valid market.
	if got := CalculatePayoutWithRake(960, 1000, 960, 500); got != 950 {
		t.Fatalf("got %d, want 950", got)
	}
}
```
</details>

File: examples/gno.land/p/demo/parimutuel/parimutuel.gno [hunk 5]
```diff
@@ -0,0 +1,77 @@
+package parimutuel
```
Comment (davd-gzl) [inline examples/gno.land/p/demo/parimutuel/parimutuel.gno:1]: Nit: the package has no `// Package parimutuel ...` comment, so its generated docs page opens on nothing. Sibling [`p/demo/svg`](https://github.com/gnolang/gno/blob/e3b7a7934/examples/gno.land/p/demo/svg/doc.gno#L2) keeps one in a `doc.gno`.

File: examples/gno.land/p/demo/parimutuel/parimutuel.gno [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+// mulDiv computes (a * b) / c. Amounts are assumed to fit comfortably
+// within int64 for realistic ugnot-denominated pools; for pools that
+// could approach 2^63, callers should pre-scale inputs or use a
+// bignum-backed alternative, which this package does not provide.
+func mulDiv(a, b, c int64) int64 {
+	if c == 0 {
+		panic("parimutuel: division by zero")
+	}
```
Comment (davd-gzl) [inline examples/gno.land/p/demo/parimutuel/parimutuel.gno]: Nit: the `c == 0` panic is unreachable. All three call sites, [`CalculatePayout`](https://github.com/gnolang/gno/blob/e3b7a7934/examples/gno.land/p/demo/parimutuel/parimutuel.gno#L12-L14), [`ImpliedProbabilityBps`](https://github.com/gnolang/gno/blob/e3b7a7934/examples/gno.land/p/demo/parimutuel/parimutuel.gno#L52-L54) and [the rake path](https://github.com/gnolang/gno/blob/e3b7a7934/examples/gno.land/p/demo/parimutuel/parimutuel.gno#L36), pass a positive `c`.

File: examples/gno.land/p/demo/parimutuel/parimutuel_test.gno [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+func TestImpliedProbabilityBpsSumsToTotal(t *testing.T) {
+	a := ImpliedProbabilityBps(500, 1000)
+	b := ImpliedProbabilityBps(300, 1000)
+	c := ImpliedProbabilityBps(200, 1000)
+	if a+b+c != 10000 {
+		t.Fatalf("expected probabilities to sum to 10000, got %d", a+b+c)
+	}
+}
```
Comment (davd-gzl) [inline examples/gno.land/p/demo/parimutuel/parimutuel_test.gno]: Nit: the name says implied probabilities sum to 10000, but every split here divides exactly. `ImpliedProbabilityBps` rounds down, so three outcomes of 1000 in a 3000 pool each return 3333 and sum to 9999.

File: examples/gno.land/p/demo/parimutuel/parimutuel.gno [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+// CalculatePayoutWithRake is CalculatePayout after deducting a commission
+// (rake) from the total pool before distribution. rakeBps is in basis
+// points: 100 = 1%, 10000 = 100%.
+//
+// If the rake consumes the entire pool (rakeBps == 10000, or rounding
+// leaves netPool at 0), there is nothing left to distribute and this
+// returns 0 rather than delegating to CalculatePayout, whose invariants
+// assume a non-degenerate pool.
+func CalculatePayoutWithRake(userBet, totalPool, winningPool, rakeBps int64) int64 {
```
Comment (davd-gzl) [inline examples/gno.land/p/demo/parimutuel/parimutuel.gno]: Suggestion: nothing exported returns the commission itself, only the payout net of it, so every realm routing the house cut re-derives `totalPool * rakeBps / 10000` inline with its own rounding.

**Comment by Gno2D2** [comment]
#### 🛠 PR Checks Summary
 🔴 Pending initial approval by a review team member, or review from tech-staff


##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)
- [ ] The pull request description provides enough details


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🔴 Pending initial approval by a review team member, or review from tech-staff


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://www.github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 The pull request was created from a fork (head branch repo: zardozmonopoly/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>Pending initial approval by a review team member, or review from tech-staff</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 Not (🔴 Pull request author is a member of the team: tech-staff)

```
### Then
```
🔴 Requirement not satisfied
└── 🔴 If
    ├── 🔴 Condition
    │   └── 🔴 Or
    │       ├── 🔴 At least one of these user(s) reviewed the pull request: [aronpark1007 davd-gzl jefft0 notJoon omarsy MikaelVallenet] (with state "APPROVED")
    │       ├── 🔴 At least 1 user(s) of the team tech-staff reviewed pull request
    │       └── 🔴 This pull request is a draft
    └── 🔴 Else
        └── 🔴 And
            ├── 🟢 This label is applied to pull request: review/triage-pending
            └── 🔴 On no pull request

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

<details><summary>The pull request description provides enough details</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 Not (🔴 Pull request author is a member of the team: core-contributors)
    └── 🟢 Not (🔴 Pull request author is user: dependabot[bot])

```
### Can be checked by
 - team core-contributors

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

**Review by notJoon — COMMENTED** [review]
It's described as a package that can be shared across realms, but I think it seems like a YAGNI for now.

File: examples/gno.land/p/demo/parimutuel/parimutuel.gno (hunk 1 above)
Comment (zardozmonopoly) [reply examples/gno.land/p/demo/parimutuel/parimutuel.gno:104]: Good catch, thanks. Switched mulDiv to overflow.Mul64p -- same helper grc20 already uses for this..

File: examples/gno.land/p/demo/parimutuel/parimutuel.gno (hunk 2 above)
Comment (zardozmonopoly) [reply examples/gno.land/p/demo/parimutuel/parimutuel.gno]: You're right, that's a real bug -- validation was against the wrong pool. Fixed to validate against totalPool before rake is applied, same as CalculatePayout. A rake reducing the payout below the winners' combined stake isn't an error condition on its own...

File: examples/gno.land/p/demo/parimutuel/parimutuel.gno (hunk 3 above)
Comment (zardozmonopoly) [reply examples/gno.land/p/demo/parimutuel/parimutuel.gno]: Yeah, that was a side effect of computing netPool before validation. Moved all the checks up front, same fix as the rake ordering issue above...

File: examples/gno.land/p/demo/parimutuel/parimutuel_test.gno (hunk 4 above)
Comment (zardozmonopoly) [reply examples/gno.land/p/demo/parimutuel/parimutuel_test.gno:10]: Added tests for both -- an overflowing pool and a winning pool above the net rake threshold...

File: examples/gno.land/p/demo/parimutuel/parimutuel.gno (hunk 5 above)
Comment (zardozmonopoly) [reply examples/gno.land/p/demo/parimutuel/parimutuel.gno:1]: Added a doc.gno, same pattern as p/demo/svg...

File: examples/gno.land/p/demo/parimutuel/parimutuel.gno (hunk 6 above)
Comment (zardozmonopoly) [reply examples/gno.land/p/demo/parimutuel/parimutuel.gno]: Fair, dropped it -- integer division by zero already panics natively and every call site guarantees c > 0 anyway...

File: examples/gno.land/p/demo/parimutuel/parimutuel_test.gno (hunk 7 above)
Comment (zardozmonopoly) [reply examples/gno.land/p/demo/parimutuel/parimutuel_test.gno]: Good point, that test data happened to divide evenly so it didn't show the rounding behavior. Renamed it to be clear it's the exact-split case and added a separate test for the more common rounding-dust case, plus a note in the doc comment...

File: examples/gno.land/p/demo/parimutuel/parimutuel.gno (hunk 8 above)
Comment (zardozmonopoly) [reply examples/gno.land/p/demo/parimutuel/parimutuel.gno]: Added CalculateRake as its own exported function for this...

**Comment by zardozmonopoly** [comment]
Fair point -- right now nothing actually imports this, so it does read as speculative. I can wire predictionmarket to use CalculatePayout/CalculatePayoutWithRake instead of its own inline math as part of this PR, which would make it a real dependency instead of a "might need it later" package. Want me to do that, or would you rather this stay standalone for now?

**Comment by notJoon** [comment]
> Fair point -- right now nothing actually imports this, so it does read as speculative. I can wire predictionmarket to use CalculatePayout/CalculatePayoutWithRake instead of its own inline math as part of this PR, which would make it a real dependency instead of a "might need it later" package. Want me to do that, or would you rather this stay standalone for now?

I think it would be better to deploy it on the testate using gnokey rather than adding it to examples (if you need this feature for your planned works). that would be the fastest route, and it would still making it available for other realms to use.

also, even if it does get added to examples, it isn't guaranteed to be included in genesis, so I think it could end up dangling package. what do you think?

**Comment by zardozmonopoly** [comment]
Fair point, and it changes my thinking here. Getting this merged into examples doesn't actually make it usable to anyone until it's part of genesis, so the "make it importable for other realms" argument for keeping this in examples mostly evaporates -- deploying it directly to a live chain via gnokey addpkg gets the same result faster and without needing a merge at all. I'll deploy it that way for what I'm building next (a bounty board realm). Not sure if that means this PR should just be closed, or if there's still value in having the source in examples as a reference even without it being live there -- happy to close it if you think it's not worth keeping open..

**Comment by notJoon** [comment]
Sorry for the late update, and thank you for all your contributions here. As discussed earlier, I'll be closing this PR.
